p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,292 @@
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		dbgId:     rand.Intn(65536),
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*webRTCStream),
+
+		acceptQueue: make(chan acceptStream, maxAcceptQueueLen),
+	}
+
+	pc.OnConnectionStateChange(conn.onConnectionStateChange)
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		if conn.IsClosed() {
+			return
+		}
+		// TODO: This seems to block on OnOpen if moved
+		// to a separate function.
+		dc.OnOpen(func() {
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Warnf("could not detach datachannel: id: %d", *dc.ID())
+				return
+			}
+			select {
+			case conn.acceptQueue <- acceptStream{rwc, dc}:
+			default:
+				pbio.NewDelimitedWriter(rwc).WriteMsg(&pb.Message{Flag: pb.Message_RESET.Enum()})
+				rwc.Close()
+			}
+		})
+	})
+
+	return conn
+}
+
+func (c *connection) resetStreams() {
+	if c.IsClosed() {
+		return
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	for k, stream := range c.streams {
+		// reset the streams, but we do not need to be notified
+		// of stream closure
+		stream.close(true, false)
+		delete(c.streams, k)
+	}
+
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{
+		Transport: "webrtc",
+	}
+}
+
+/* Close closes the underlying peerconnection.
+ */
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	return c.pc.Close()
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	dc, err := c.pc.CreateDataChannel("", nil)
+	if err != nil {
+		return nil, err
+	}
+	rwc, err := c.detachChannel(ctx, dc)
+	if err != nil {
+		return nil, fmt.Errorf("could not open stream: %w", err)
+	}
+	stream := newStream(
+		c,
+		dc,
+		rwc,
+		nil,
+		nil,
+	)
+	c.addStream(stream)
+	return stream, nil
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case accStream := <-c.acceptQueue:
+		stream := newStream(
+			c,
+			accStream.channel,
+			accStream.stream,
+			nil,
+			nil,
+		)
+		c.addStream(stream)
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+// only used during setup
+func (c *connection) setRemotePeer(id peer.ID) {
```

Comment:
do you really need to have all these setters the runtime object? it makes it a lot more complex to reason about the object. I imagine you only want to set this at construction time, so why not just assign this from the constructor directly and leave it at that. you could also opt for one of many factory-like styles and have a separate constructor object from the actual runtime (created) object.

Just saying as now a true in-depth reviewer also has to keep in mind that I need to check all places where this object is mutated through these setters.

Reply:
I understand the issue and it has been raised earlier too, but the problem here is as follows:
Consider a peer A establishing a connection to B.
Connection establishment requires a noise handshake XX to be completed from B to A. With reference to the handshake pattern described [here](https://noiseexplorer.com/patterns/XX/), we can see that the initiator knows the noise handshake is complete before the server. This means that the noise initiator (B) can now start opening datachannels on the peerconnection, before the remote is aware that the handshake succeeded.

The `connection` struct is responsible for setting the callbacks for handling incoming datachannels. This is because we need to wrap them and queue them until `AcceptStream` is called on the connection. Consider the pion handling of incoming channels [here](https://github.com/pion/webrtc/blob/602cedd4fe8d7a49c4d3526c720d0be3056af521/sctptransport.go#L282-L303), the datachannel will be discarded unless there is a handler present for the datachannel. This leads to a problem where a stream is initiated by a peer and not handled by the remote peer, leading to reads timing out. Please note that changing which party initiates the noise handshake does not solve this problem, since there will always be a peer which knows first that the handshake is complete.

Since we cannot delay or stop the peerconnection from handling datachannels, the solution used here is to wrap the peerconnection in the connection object before starting the noise handshake. The drawback for B is that it cannot know the id and public key of the remote peer until the handshake is complete. Therefore, I had to add a setter for the id and public key after initiation of the noise handshake.

Reply:
Given the implications at the very least you'll want to add this information to the code as well in that case :) so the intention is crystal clear. Whether there is no solution beyond a setter is a different question, as I can offer some ideas regarding that. But for now, I imagine given how the low severity this has, let's leave it at this :) Feel free to resolve as is!